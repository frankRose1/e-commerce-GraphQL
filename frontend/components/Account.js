import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import {formatDistance} from 'date-fns';
import styled from 'styled-components';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import OrderItem from './styles/OrderItemStyles';
import Error from './ErrorMessage';
import Loading from './Loading';
import SubTitle from './styles/SubTitleStyles';

const OrdersGrid = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    userOrders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        quantity
        description
        title
        price
        image
      }
    }
  }
`;

class Account extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({data, loading, error}) => {
          const {userOrders} = data;
          if (loading) return <Loading />
          if(error) return <Error error={error} />
          if (!userOrders.length) return <p>Looks like you havn't ordered anything yet!</p>
          
          return (
            <div>
              <SubTitle>Your Account</SubTitle>
              <h2>You have {userOrders.length} order{userOrders.length === 1 ? '' : 's'}</h2>
              <OrdersGrid>
                {userOrders.map(order => (
                  <OrderItem key={order.id}>
                    <Link href={{
                      pathname: '/order',
                      query: { orderId: order.id }
                    }}>
                      <a>
                        <div className="order-meta">
                          <p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
                          <p>{order.items.length} Product{order.items.length === 1 ? '' : 's'}</p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)}</p>
                        </div>
                        <div className="images">
                          {order.items.map(item => (
                            <img 
                              src={item.image} 
                              alt={item.title}
                              key={item.id} />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItem>
                ))}
            </OrdersGrid>
            </div>
          )
        }}
      </Query>
    );
  }
}

export default Account;