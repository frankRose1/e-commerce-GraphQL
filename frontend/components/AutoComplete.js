import React, { Component } from 'react';
import {ApolloConsumer} from 'react-apollo'; //run queries on demand
import gql from 'graphql-tag';
import Router from 'next/router';
import debounce from 'lodash.debounce';
import Downshift, {resetIdCounter} from 'downshift';
import {DropDown, DropDownItem, SearchStyles} from './styles/DropDown';

function routeToItem(item){
  Router.push({
    pathname: '/item',
    query: { id: item.id }
  })
}

//serach by either the title or description
const SEARCH_ITEMS_QUERY = gql`
  query SEARCH_ITEMS_QUERY($searchTerm: String!){
    items(where: {
      OR: [
        { title_contains: $searchTerm},
        { description_contains: $searchTerm}
      ]
    }){
      id
      image
      title
    }
  }
`;

class AutoComplete extends Component {

  state= {
    loading: false,
    items: []
  }

  //Manually fire off a query on input change
  handleChange = debounce(async (e, {query}) => {
    this.setState({loading: true});
    const res = await query({
      query: SEARCH_ITEMS_QUERY,
      variables: { searchTerm: e.target.value }
    });
    this.setState({
      loading: false,
      items: res.data.items
    })
  }, 400);

  render() {
    const {items, loading} = this.state;
    resetIdCounter();
    return (
      <SearchStyles>
        <Downshift
          onChange={routeToItem}
          itemToString={item => (item === null ? '' : item.title)}>
          { ({getInputProps, getItemProps, isOpen, inputValue, highlightedIndex}) => (
          <div>
            <ApolloConsumer>
              { client => (
                <input
                  {...getInputProps({
                    placeholder: "Search Our Awesome Collection",
                    type: "search",
                    id: 'search',
                    className: this.state.loading ? 'loading' : '',
                    onChange: e => {
                      e.persist();
                      this.handleChange(e, client);
                    }
                  })} />
              )}
            </ApolloConsumer>
            {isOpen && (
              <DropDown>
                {items.map((item, i) => (
                  <DropDownItem
                    {...getItemProps({item})}
                    highlighted={highlightedIndex === i}
                    key={item.id}>
                    <img width="50" src={item.image} alt={item.title}/>
                    {item.title}
                  </DropDownItem>
                ))}
                {!items.length && !loading && (
                  <DropDownItem>
                    Nothing found for "{inputValue}"
                  </DropDownItem>
                )}
              </DropDown>
            )}
          </div>
          )}
        </Downshift>
      </SearchStyles>
    );
  }
}

export default AutoComplete;