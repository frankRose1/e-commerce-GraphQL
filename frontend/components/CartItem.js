import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: auto 1fr auto;
  align-items: center;
  img {
    width: 100px;
    margin-right: 10px;
  }
  h3, p {
    margin: 0;
  }
`;

const DeletedItem = styled.li`
  padding: 1rem 0;
  display: grid;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  grid-template-columns: 1fr auto;
  align-items: center;
`;

const CartItem = ({cartItem: { item, quantity, id} }) => {
  //in the case that An item is deleted in the DB we dont want to show it in the users cart
  if (!item){
    return (
      <DeletedItem>
        <p>Sorry, this item is no longer sold in our store!</p>
        <RemoveFromCart cartItemId={id}/>
      </DeletedItem>
    )
  }

  return (
  <CartItemStyles>
    <img src={item.image} alt={item.title}/>
    <div className="cart-item-details">
      <h3>{item.title}</h3>
      <p>
        {formatMoney(item.price * quantity)}
        {' - '}
        <em>{quantity} &times; {formatMoney(item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart cartItemId={id}/>
  </CartItemStyles>
)};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    quantity: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired,
    item: PropTypes.object
  }).isRequired
};

export default CartItem;