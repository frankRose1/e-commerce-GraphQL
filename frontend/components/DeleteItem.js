import React, { Component } from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import {ALL_ITEMS_QUERY} from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!){
    deleteItem(id: $id){
      id
    }
  }
`;

class DeleteItem extends Component {

  static propTypes = {
    id: PropTypes.string.isRequired
  };

  //maually update the cache so that it matches the data on the server. and updating cache will update the UI
  updateCache = (cache, payload) => {
    //read the cache for the items. we need to do this wit the query used to fetch the items in the first place
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // remove the deleted ID from cache
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    //put the items back
    cache.writeQuery({query: ALL_ITEMS_QUERY, data});
  }

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{
          id: this.props.id
        }}
        update={this.updateCache}>
        {(deleteItem, {error, loading}) => (
          <button
          disabled={loading}
          onClick={() => {
            if (confirm('Are you sure you want to delete this item?')){
              deleteItem().catch(err => {
                alert(err.message);
              });
            }
          }}>
            Delete This Item
          </button>
        )}
      </Mutation>
    );
  }
}



export default DeleteItem;