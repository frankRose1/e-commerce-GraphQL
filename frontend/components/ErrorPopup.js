import React from 'react';
import PropTypes from 'prop-types';
import ErrPopUpStyles from './styles/ErrPopUpStyles';

//meant to handle errors on buttons where the UI is too small to display an error message

const ErrorPopup = ({error}) => {
  
  console.log(error);
  return (
    <ErrPopUpStyles showPopup={error ? true : false}>
      <p>{error}</p>
    </ErrPopUpStyles>
  );
};

ErrorPopup.propTypes = {
  error: PropTypes.string.isRequired,
};

export default ErrorPopup;