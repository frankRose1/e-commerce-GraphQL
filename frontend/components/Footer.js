import React from 'react';
import { FooterStyles, FooterLogo } from './styles/FooterStyles';
import Link from 'next/link';

const Footer = () => (
  <FooterStyles>
    <FooterLogo>BackPackzz</FooterLogo>
    <div className='footer-cols'>
      <div>
        <h4>Around The Web</h4>
        <ul>
          <li>
            <a href='https://www.facebook.com/' target='_blank'>
              {' '}
              <i className='fab fa-facebook-f' /> facebook
            </a>
          </li>
          <li>
            <a href='https://www.instagram.com/' target='_blank'>
              <i className='fab fa-instagram' /> instagram
            </a>
          </li>
          <li>
            <a href='https://www.twitter.com/' target='_blank'>
              <i className='fab fa-twitter' /> twitter
            </a>
          </li>
        </ul>
      </div>
      <div>
        <h4>Company</h4>
        <ul>
          <li>
            <Link href='/'>
              <a>Home</a>
            </Link>
          </li>
          <li>
            <Link href='/shop'>
              <a>Shop</a>
            </Link>
          </li>
          <li>
            <Link href='/'>
              <a>About Us</a>
            </Link>
          </li>
          <li>
            <Link href='/'>
              <a>Customer Service</a>
            </Link>
          </li>
        </ul>
      </div>
      <div>
        <h4>NewsLetter</h4>
        <p>
          Subscribe to get special offers, free giveaways, and
          once-in-a-lifetime deals!(psst! this is just a placeholder!)
        </p>
        <div className='subscribe'>
          <input type='email' placeholder='your-email@example.com' />
          <input type='submit' value='SUBSCRIBE' />
        </div>
      </div>
    </div>
    <p>&copy; BackPackzz {new Date().getFullYear()} | All rights Reserved</p>
  </FooterStyles>
);

export default Footer;
