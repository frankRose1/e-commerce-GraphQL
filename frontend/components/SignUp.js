import React, { Component } from 'react';
import gql from 'graphql-tag';
import {Mutation} from 'react-apollo';
import {CURRENT_USER_QUERY} from './User';
import Error from './ErrorMessage';
import Form from './styles/Form';

const SIGN_UP_MUTATION = gql`
  mutation SIGN_UP_MUTATION($email: String!, $name: String!, $password: String!, $confirmPassword: String!){
    signup(email: $email, name: $name, password: $password, confirmPassword: $confirmPassword){
      id
    }
  }
`;

class SignUp extends Component {
  state = {
    name: '',
    email: '',
    password: '',
    confirmPassword: ''
  }

  handleInputChange = e => {
    const {value, name} = e.target;
    this.setState({ [name]: value });
  }

  render() {
    return (
      <Mutation
        mutation={SIGN_UP_MUTATION}
        variables={this.state}
        refetchQueries={[ {query: CURRENT_USER_QUERY} ]}>
        {(signup, {error, loading}) => (
          <Form 
            data-test="signup-form"
            method="post" 
            onSubmit={async e => {
            e.preventDefault();
            await signup();
            this.setState({ email: '', password: '', confirmPassword: '', name: ''})
          }}>
          
          <fieldset disabled={loading} aria-busy={loading}>
            <Error error={error}/>
            <h2>Sign Up</h2>

            <label htmlFor="name">
              Name
              <input 
                type="text" 
                id="name"
                name="name"
                placeholder="Name"
                value={this.state.name}
                onChange={this.handleInputChange}
                required/>
            </label>

            <label htmlFor="email">
              Email Address
              <input 
                type="email" 
                id="email"
                name="email"
                placeholder="Email"
                value={this.state.email}
                onChange={this.handleInputChange}
                required/>
            </label>

            <label htmlFor="password">
              Password
              <input 
                type="password" 
                id="password"
                name="password"
                placeholder="Password"
                value={this.state.password}
                onChange={this.handleInputChange}
                required/>
            </label>

            <label htmlFor="confirmPassword">
              Confirm Password
              <input 
                type="password" 
                id="confirmPassword"
                name="confirmPassword"
                placeholder="Confirm Password"
                value={this.state.confirmPassword}
                onChange={this.handleInputChange}
                required/>
            </label>

            <button 
              type="submit"
              disabled={loading}>
              Sig{loading ? 'ning': 'n'} Up!
            </button>
          </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default SignUp;
export {SIGN_UP_MUTATION};