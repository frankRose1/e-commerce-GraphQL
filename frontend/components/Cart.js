import React from 'react';
import {Mutation, Query} from 'react-apollo';
import gql from 'graphql-tag';
import {adopt} from 'react-adopt';
import User from './User';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import Button from './styles/SickButton';
import PurchaseItem from './PurchaseItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

//Client side query & mutation for cart boolean
const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({render}) => <User>{render}</User>,
  toggleCart: ({render}) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({render}) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => {
  return (
    <Composed>
      {( {user, toggleCart, localState} ) => {
        const {data: {me}, loading} = user;
        if (loading) return <p>Loading...</p>
        if(!me) return null //dont show the cart if user is not signed in
        return (
          <CartStyles
            data-test="cart"
            open={localState.data.cartOpen}>
            <header>
              <CloseButton onClick={toggleCart} title="close">&times;</CloseButton>
              <Supreme>{me.name}'s Cart</Supreme>
              <p>{me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart</p>
            </header>

              <ul>
                {me.cart.map(cartItem => (
                  <CartItem 
                    key={cartItem.id}
                    cartItem={cartItem}/>
                ))}
              </ul>

            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              {me.cart.length
                ? (
                  <PurchaseItem>
                    <Button>Checkout</Button>
                  </PurchaseItem>
                )
              : null }
            </footer>
          </CartStyles>
        )
      }}
    </Composed>
  );
};

export default Cart;
export {TOGGLE_CART_MUTATION, LOCAL_STATE_QUERY};