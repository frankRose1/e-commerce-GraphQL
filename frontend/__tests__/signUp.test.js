import {mount} from 'enzyme';
import {MockedProvider} from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import {ApolloConsumer} from 'react-apollo';
import wait from 'waait';
import SignUp, {SIGN_UP_MUTATION} from '../components/SignUp';
import {CURRENT_USER_QUERY} from '../components/User';
import {fakeUser} from '../lib/testUtils';

function type(wrapper, name, value){
  wrapper.find(`input[name="${name}"]`).simulate('change', {target: {name, value} });
}

const me = fakeUser();
const mocks = [
  {
    request: {
      query: SIGN_UP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password: 'test123',
        confirmPassword: 'test123'
      }
    },
    result: {
      data: {
        signup: {
          __typename: 'User',
          name: me.name,
          email: me.email,
          id: 'abc123'
        }
      }
    },
  },
  //current user query mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me } }
  }
];

/**
 * After a user signs up, querying the current user should return data about the user
 * this is how to check that both the mutation and query are working
 */

describe('<SignUp />', () => {

  it('renders and matches the snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="signup-form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  //after the signup mutation, the current user query is called. need to apollo client to call the query manually
  it('calls the signup mutation properly', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <SignUp />
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    //simulate typing in to the form
    type(wrapper, 'name', me.name);
    type(wrapper, 'email', me.email);
    type(wrapper, 'password', 'test123');
    type(wrapper, 'confirmPassword', 'test123');
    wrapper.update()
    wrapper.find('form[data-test="signup-form"]').simulate('submit');
    await wait(30);
    //query the user out of the apollo client
    const user = await apolloClient.query({query: CURRENT_USER_QUERY });
    expect(user.data.me).toMatchObject(me);
  });
});