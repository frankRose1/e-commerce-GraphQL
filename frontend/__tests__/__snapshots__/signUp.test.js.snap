// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches the snapshot 1`] = `
<form
  className="Form-sc-14226r0-0 kzFose"
  data-test="signup-form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <DisplayError
      error={Object {}}
    />
    <h2>
      Sign Up
    </h2>
    <label
      htmlFor="name"
    >
      Name
      <input
        id="name"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="email"
    >
      Email Address
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="confirmPassword"
    >
      Confirm Password
      <input
        id="confirmPassword"
        name="confirmPassword"
        onChange={[Function]}
        placeholder="Confirm Password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <button
      disabled={false}
      type="submit"
    >
      Sig
      n
       Up!
    </button>
  </fieldset>
</form>
`;
